compiler/src/dotty/tools/dotc/transform/TypeTestsCasts.scala

```diff
@@ -358,7 +358,11 @@ object TypeTestsCasts {
                 report.error(em"$untestable cannot be used in runtime type tests", tree.srcPos)
                 constant(expr, Literal(Constant(false)))
               case _ =>
-                val erasedTestType = erasure(testType)
+                val erasedTestType =
+                  if testType.isAny && expr.tpe.isPrimitiveValueType then
+                    defn.AnyValType
```

Comment:
Isn't that going to trigger with the same issue with any *other* type that erases to `Object`? Like ... `AnyVal` itself?

There's supposed to already be logic in `transformIsInstanceOf` to avoid a type test when `expr.tpe <:< testType`:
https://github.com/scala/scala3/blob/9961e707a65ca9f8adb0128071bbecad43fa0a7e/compiler/src/dotty/tools/dotc/transform/TypeTestsCasts.scala#L259-L261
The problem is that it tests on erased types, instead of the original types. Addressing that at the root should be more robust, shouldn't it?